FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt

```diff
@@ -65,4 +81,198 @@ object PgpMsg {
+    }
+
+    val isArmored = chunk.contains(ARMOR_HEADER_DICT[MsgBlock.Type.ENCRYPTED_MSG]!!.begin)
+    if (isArmored) input = ArmoredInputStream(input)
+
+    val protector = if (passphrase != null) {
+      PasswordBasedSecretKeyRingProtector.forKey(pgpSecretKeyRing, passphrase)
+    } else {
+      UnprotectedKeysProtector()
+    }
+
+    try {
+      val d = PGPainless.decryptAndOrVerify()
+          .onInputStream(input)
+          .decryptWith(protector, PGPSecretKeyRingCollection(listOf(pgpSecretKeyRing)))
+
+      val decryptionStream = if (pgpPublicKeyRingCollection != null) {
+        d.verifyWith(pgpPublicKeyRingCollection).ignoreMissingPublicKeys().build()
+      } else {
+        d.doNotVerify().build()
+      }
+
+      val output = ByteArrayOutputStream()
+      decryptionStream.use { it.copyTo(output) }
+
+      val streamResult = decryptionStream.result
+      return DecryptionResult.withDecrypted(output, streamResult.fileInfo?.fileName)
+    } catch (ex: MessageNotIntegrityProtectedException) {
+      return DecryptionResult(
+        content = null,
+        isEncrypted = true,
+        filename = null,
+        signature = null,
+        error = DecryptionError(
+          DecryptionErrorType.NO_MDC,
+          "Security threat! Message is missing integrity checks (MDC). The sender should " +
+          "update their outdated software."
+        )
+      )
+    } catch (ex: ModificationDetectionException) {
+      return DecryptionResult(
+          content = null,
+          isEncrypted = true,
+          filename = null,
+          signature = null,
+          error = DecryptionError(
+              DecryptionErrorType.BAD_MDC,
+              "Security threat! Integrity check failed."
+          )
+      )
+    } catch (ex: Exception) {
+      if (ex is PGPDataValidationException || (ex is PGPException
+              && (
+                ex.message?.contains("exception decrypting session info") == true
+                || ex.message?.contains("encoded length out of range") == true
+                || ex.message?.contains("Invalid Curve25519 public key") == true
+                || ex.message?.contains("Exception recovering session info") == true
+                || ex.message?.contains("ECDH requires use of PGPPrivateKey for decryption") == true
+              )
+          )
+      ) {
+        return if (passphrase != null) {
+          DecryptionResult.withError(
+              type = DecryptionErrorType.WRONG_PASSWORD,
+              message = "Wrong password",
+              cause = ex
+          )
+        } else {
+          DecryptionResult.withError(
+              type = DecryptionErrorType.KEY_MISMATCH,
+              message = "There is no matching key",
+              cause = ex
+          )
+        }
+      } else {
+        return DecryptionResult.withError(
+            type = DecryptionErrorType.OTHER,
+            message = "There is no matching key",
+            cause = ex
+        )
```

Comment:
This is also naive - this could be a whole range of other issues.

Reply:
will change to something more general like "decryption failed"